internal/controller/util/json_util.go

```diff
@@ -7,78 +7,278 @@ import (
-	"k8s.io/apimachinery/pkg/types"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
 	"k8s.io/client-go/util/jsonpath"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 const (
 	defaultTimeoutValue       = 300
-	pollInterval              = 100
 	expectedNumberOfJSONPaths = 2
+	podType                   = "pod"
+	deploymentType            = "deployment"
+	statefulsetType           = "statefulset"
+	pInterval                 = 100
 )
 
-func getJSONObject(k8sClient client.Client, hook *kubeobjects.HookSpec) (client.Object, error) {
-	var obj client.Object
-
-	switch hook.SelectResource {
-	case "pod":
-		obj = &corev1.Pod{}
-	case "deployment":
-		obj = &appsv1.Deployment{}
-	case "statefulset":
-		obj = &appsv1.StatefulSet{}
-	default:
-		return obj, fmt.Errorf("unsupported resource type %s", hook.SelectResource)
+func EvaluateCheckHook(k8sClient client.Client, hook *kubeobjects.HookSpec, log logr.Logger) (bool, error) {
+	if hook.LabelSelector == nil && hook.NameSelector == "" {
+		return false, fmt.Errorf("either nameSelector or labelSelector should be provided to get resources")
 	}
 
-	err := k8sClient.Get(context.Background(),
-		types.NamespacedName{Name: hook.NameSelector, Namespace: hook.Namespace},
-		obj)
-
-	return obj, err
-}
-
-func EvaluateCheckHook(k8sClient client.Client, hook *kubeobjects.HookSpec, log logr.Logger) (bool, error) {
 	timeout := getTimeoutValue(hook)
 
+	pollInterval := pInterval * time.Microsecond
+
 	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
 	defer cancel()
 
-	ticker := time.NewTicker(pollInterval * time.Millisecond)
+	ticker := time.NewTicker(pollInterval)
 	defer ticker.Stop()
 
-	for {
+	for int(pollInterval.Seconds()) < timeout {
 		select {
 		case <-ctx.Done():
 			return false, fmt.Errorf("timeout waiting for resource %s to be ready: %w", hook.NameSelector, ctx.Err())
 		case <-ticker.C:
-			obj, err := getJSONObject(k8sClient, hook)
+			objs, err := getResourcesList(k8sClient, hook)
 			if err != nil {
-				if k8serrors.IsNotFound(err) {
-					// Resource not found, continue polling
-					continue
-				}
-
 				return false, err // Some other error occurred, return it
 			}
 
-			res, err := EvaluateCheckHookExp(hook.Chk.Condition, obj)
-			if err != nil {
-				continue // This may mean that expression is not evaluated
+			if len(objs) == 0 {
+				pollInterval *= 2
+				ticker.Reset(pollInterval)
+
+				continue
 			}
 
-			return res, nil
+			return evaluateCheckHookForObjects(objs, hook, log)
+		}
+	}
+
+	return false, nil
+}
+
+func evaluateCheckHookForObjects(objs []client.Object, hook *kubeobjects.HookSpec, log logr.Logger) (bool, error) {
+	finalRes := true
+
+	var errOut error
+
+	for _, obj := range objs {
+		res, err := EvaluateCheckHookExp(hook.Chk.Condition, obj)
+		finalRes = finalRes && res
+
+		if err != nil {
+			log.Info("error executing check hook", "for", hook.Name, "with error", err)
+			errOut = err
```

Comment:
@asn1809 , when we have real failure here, I mean not res = false, but err is not empty, we need to proceed in this loop? 
Once finalRes becomes false, no matter how many iterations follow, it will remain false

Reply:
Thanks @ELENAGER, the reason is to fail the hook if any of the checks fail. For example, suppose a labelSelector or nameSelector returns 2 Deployments, and check hook passes for one and fails for other, then overall check hook should be marked as failed. Let me know if you see any issues.

Reply:
@Annaraya-Narasagond , exactly, maybe just 
`return false, err`
here?